### ISSUE TITLE
Expose profile name to build.rs / program at runtime

### ISSUE THREAD

**Issue by banool** [rust-lang/cargo#11054]
### Problem

Today it is very difficult to get the build profile name, and perhaps impossible to do accurately ([see here](https://stackoverflow.com/questions/73595435/how-to-get-profile-from-cargo-toml-in-build-rs-or-at-runtime/73603419#73603419)) to get the profile name, e.g. for logging purposes.

### Proposed Solution

It would be great if the profile name were exposed to `build.rs` / a Rust program at runtime. There is a `PROFILE` env var already but it only contains either `release` or `dev`. I'd suggest we add something new like `BUILD_PROFILE_NAME` that contains the actual name of the profile used to build the binary.

In other words, it'd be great if this code worked:
```
cargo run --profile bench
```
```
println!("BUILD_PROFILE_NAME: {}", std::env!("BUILD_PROFILE_NAME"));
```
```
BUILD_PROFILE_NAME: bench
```

And this:
```
cargo run --profile my-custom-profile
```
```
println!("BUILD_PROFILE_NAME: {}", std::env!("BUILD_PROFILE_NAME"));
```
```
BUILD_PROFILE_NAME: my-custom-profile
```

### Notes

Depending on how this is implemented, we might need to address https://github.com/rust-lang/cargo/issues/11053 first.

**Comment by ehuss** [comment]
Can you say more about why you want the profile name?

**Comment by banool** [comment]
We had a recent issue at my company where it became relevant what build profile our binary was using, and while we had a good guess, it would've been much easier if the binary just logged it at startup so we could be certain.

**Comment by oerdn** [comment]
> We had a recent issue at my company where it became relevant what build profile our binary was using, and while we had a good guess, it would've been much easier if the binary just logged it at startup so we could be certain. If this value was exposed somehow, we could do that logging.

I am not sure this is a good practice or not;  values for custom profile and build scripts can change over time and this will break the synchronization between your existing binary and the current profile state. What I mean is, this logging may mislead you, it is hard to track the changes, you can still log values with the profile name but this brings a question what is the real need of the identifier then?

**Comment by oerdn** [comment]
> As part of this startup logging we log the commit hash the binary was built from, so we can easily see what the config was at that time. I'm not sure how else this information could mislead us. So the need is like I said, we had a production issue where it became relevant what profile was used to build the binary, and we wish we knew at runtime from startup logs.

Well, my intention was to find some reliable(by itself) generic solution to a generic problem(currently unknown), everything here is so specific. I am not opposed the idea btw, just tried to support, fingers crossed ^^

**Comment by nirvana-msu** [comment]
My use-case for knowing Cargo profile at either runtime or compile time:  I'd like to set higher I/O timeouts in my code when profiling with Valgrind, as otherwise connections/requests always timeout due to a massive slowdown introduced by Valgrind.

I have a custom Cargo profile to use with Valgrand (inheriting from release and enabling debug info). I don't want to change timeouts for normal debug/release builds, hence no way to differentiate other than by Cargo profile name.

**Comment by dlrobson** [comment]
My team uses the build.rs file to configure a C++ cmake project using cmake-rs. We link against the generated library using cxx-rs. We also run thread/address sanitizers on the C++ code. I want to be able to read what profile was specified so that we can apply the appropriate cmake flags for the sanitizers.

**Comment by JSorngard** [comment]
Another use-case: in one of my crates I use the `no-panic` crate to verify that my crate has no panics. It would be unnecessary for the end user to have my crate depend on `no-panic` though, so I use a build script to check an environment variable and enable a `cfg` that enables the panic check.
However, that panic check only works correctly if the test binary is compiled in a profile with fat LTO. I have a profile for that, but it would be nice if there was a robust way to verify that the crate is currently being compiled with that profile so that I can offer up a useful warning if it is not.

**Comment by epage** [comment]
From a recent [zulip thread](https://rust-lang.zulipchat.com/#narrow/channel/246057-t-cargo/topic/Way.20to.20get.20profile.20config/near/495834436)

@epage 

> I suspect it would be better to focus on the features of a profile, rather than a profile name, so people are free to inherit and extend a profile.
>
> I thought I've written on that lately but not seeing it in that issue

@ehuss 

> Yea, when we stabilized custom named profiles, we decided that we didn't want to expose the profile name to build scripts for a few reasons. However, exposing the profile setting might be feasible if there is a use-case. We just have to be aware that package overrides can make that a little more complicated.

Based on that context, I'm going to close this.  As mentioned, if there are profile settings people want exposed, you are welcome to open issues for them to be discussed and designed.
